### PR TITLE
[#174294099] Migrate LinkedIn OAuth to communitarian

### PR DESCRIPTION
**Pull request by Vadim-Kurnatovskiy**
[Story 1](https://www.pivotaltracker.com/story/show/174294099), [Story 2](https://www.pivotaltracker.com/story/show/174294181)

### CODE REVIEW

File: lib/auth/linkedin_authenticator.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    result.extra_data = {
+      uid: oauth2_uid,
+      provider: oauth2_provider
+    }
+
+    username_max_length = SiteSetting.max_username_length.to_i
+    username_id = result.user ? result.user.id : Time.current.to_i
+    result.username = username = [data[:first_name], data[:last_name], username_id].join(".")[0, username_max_length]
+
+    if result.user && email && (result.user.email != email)
```
Comment (DmitryBarskov) [inline lib/auth/linkedin_authenticator.rb]:
```suggestion
    if email && (result.user&.email != email)
```

File: lib/auth/linkedin_authenticator.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    result.user = oauth2_user_info.try(:user)
+    result.email_valid = @opts[:trusted]
+
+    result.extra_data = {
+      uid: oauth2_uid,
+      provider: oauth2_provider
+    }
+
+    username_max_length = SiteSetting.max_username_length.to_i
```
Comment (DmitryBarskov) [inline lib/auth/linkedin_authenticator.rb]: I think it should be a method

File: lib/auth/linkedin_authenticator.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    username_id = result.user ? result.user.id : Time.current.to_i
```
Comment (DmitryBarskov) [inline lib/auth/linkedin_authenticator.rb]:
```suggestion
    username_id = result.user&.id || Time.current.to_i
```

File: lib/auth/linkedin_authenticator.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    result.username = username = [data[:first_name], data[:last_name], username_id].join(".")[0, username_max_length]
```
Comment (DmitryBarskov) [inline lib/auth/linkedin_authenticator.rb]: Could you create a method for that?

File: lib/auth/linkedin_authenticator.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    if result.user && email && (result.user.email != email)
+      begin
+        result.user.primary_email.update!(email: email)
+        result.user.username.update!(username: username)
+      rescue
```
Comment (DmitryBarskov) [inline lib/auth/linkedin_authenticator.rb]: Only updating of email can fail here?

File: lib/auth/linkedin_authenticator.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  end
+
+  def after_authenticate(auth_token)
+    result = Auth::Result.new
+
+    oauth2_provider = auth_token[:provider]
+    oauth2_uid = auth_token[:uid]
+    data = auth_token[:info]
+    result.email = email = data[:email]
+    result.name = name = [data[:first_name], data[:last_name]].join(" ")
```
Comment (DmitryBarskov) [inline lib/auth/linkedin_authenticator.rb:19]: Do you like multiple assignment?

File: lib/auth/linkedin_authenticator.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    username_max_length = SiteSetting.max_username_length.to_i
+    username_id = result.user ? result.user.id : Time.current.to_i
+    result.username = username = [data[:first_name], data[:last_name], username_id].join(".")[0, username_max_length]
+
+    if result.user && email && (result.user.email != email)
+      begin
+        result.user.primary_email.update!(email: email)
+        result.user.username.update!(username: username)
```
Comment (DmitryBarskov) [inline lib/auth/linkedin_authenticator.rb]: A question: should this block be executed in a transaction?

File: assets/javascripts/initializers/communitarian.js.es6 [hunk 8]
```diff
@@ -10,6 +10,7 @@ import { extractError } from "discourse/lib/ajax-error";
 function initializeCommunitarian(api) {
   api.modifyClass("controller:create-account", {
     performAccountCreation() {
+      return this._super(...arguments) if(this.get("authOptions.email") == this.accountEmail);
```
Comment (DmitryBarskov) [inline assets/javascripts/initializers/communitarian.js.es6]: Won't work in JS 😄

Comment (DmitryBarskov) [inline assets/javascripts/initializers/communitarian.js.es6]:
```suggestion
      if (this.get("authOptions.email") == this.accountEmail) {
        return this._super(...arguments);
      }
```

File: lib/auth/linkedin_authenticator.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        Rails.logger.warn("FAILED to update email for #{user.username} to #{email} cause it is in use by #{used_by}")
+      end
+    end
+  end
+
+  def update_username(result, username)
+    if username && (result.user.username != username)
+      begin
+        result.user.username.update!(username: username)
+      rescue
```
Comment (DmitryBarskov) [inline lib/auth/linkedin_authenticator.rb:80]: I suggest to rescue from `ActiveRecord::RecordInvalid` only

Comment (Vadim-Kurnatovskiy) [reply lib/auth/linkedin_authenticator.rb:80]: I think there may be various errors

**Review by DmitryBarskov — APPROVED** [review]

**Review by ShaidullinArtur — APPROVED** [review]

**Pull request merged**